ntp-udp/src/lib.rs

```diff
@@ -160,16 +163,15 @@ impl UdpSocket {
     async fn fetch_send_timestamp(&self, expected_counter: u32) -> io::Result<NtpTimestamp> {
         trace!("waiting for timestamp socket to become readable to fetch a send timestamp");
         loop {
-            // here we wait for the socket to become writable again, even though what we want to do
-            // is read from the error queue.
+            // Send timestamps are sent to the udp socket's error queue. Sadly, tokio does not
+            // currently support awaiting whether there is something in the error queue
+            // see https://github.com/tokio-rs/tokio/issues/4885.
             //
-            // We found that waiting for `readable()` is unreliable, and does not seem to actually
-            // fire when the timestamping message is available. To our understanding, the socked
-            // becomes `writable()` when it has sent all its packets. So in practice this is also
-            // the moment when the timestamp message is available in the error queue.
-            let mut guard = self.io.writable().await?;
-            match guard.try_io(|inner| fetch_send_timestamp_help(inner.get_ref(), expected_counter))
-            {
+            // Therefore, we manually configure an extra file descriptor to listen for POLLPRI on
+            // the main udp socket. This `exceptional_condition` file descriptor becomes readable
+            // when there is something in the error queue.
+            let mut guard = self.exceptional_condition.readable().await?;
+            match guard.try_io(|_| fetch_send_timestamp_help(self.io.get_ref(), expected_counter)) {
```

Comment:
here we use a guard on the `exceptional_condition` to perform io on the main `io` udp socket. I'm not 100% sure that this is correct if the `io` udp socket were to e.g. block.

Reply:
From how I understand how tokio works internally, this is the correct approach here.